Don't load source map until the JS source is fully received

b-fuze opened the pull request:
We have a Node.js script that is writing the source map while also piping the Javascript source from Babel into UglifyJS. The issue is that UglifyJS will try to load the source map right when it starts up and is parsing its arguments, this PR delays that until the source(s) have completely transferred via piping or other means

alexlamsl commented:
Something like this could work:
```diff
--- a/bin/uglifyjs
+++ b/bin/uglifyjs
@@ -46,7 +46,7 @@ program.option("--name-cache <file>", "File to hold mangled name mappings.");
 program.option("--rename", "Force symbol expansion.");
 program.option("--no-rename", "Disable symbol expansion.");
 program.option("--self", "Build UglifyJS as a library (implies --wrap UglifyJS)");
-program.option("--source-map [options]", "Enable source map/specify source map options.", parse_source_map());
+program.option("--source-map [options]", "Enable source map/specify source map options.", parse_js());
 program.option("--timings", "Display operations run time on STDERR.")
 program.option("--toplevel", "Compress and/or mangle variables in toplevel scope.");
 program.option("--verbose", "Print diagnostic messages.");
@@ -176,6 +176,11 @@ function run() {
     UglifyJS.AST_Node.warn_function = function(msg) {
         print_error("WARN: " + msg);
     };
+    var content = program.sourceMap && program.sourceMap.content;
+    if (content && content != "inline") {
+        print_error("INFO: Using input source map: " + content);
+        options.sourceMap.content = read_file(content, content);
+    }
     if (program.timings) options.timings = true;
     try {
         if (program.parse) {
@@ -377,19 +382,6 @@ function parse_js(flag) {
     }
 }

-function parse_source_map() {
-    var parse = parse_js();
-    return function(value, options) {
-        var hasContent = options && "content" in options;
-        var settings = parse(value, options);
-        if (!hasContent && settings.content && settings.content != "inline") {
-            print_error("INFO: Using input source map: " + settings.content);
-            settings.content = read_file(settings.content, settings.content);
-        }
-        return settings;
-    }
-}
-
 function skip_key(key) {
     return skip_keys.indexOf(key) >= 0;
 }
```
All that's left is a `mocha` test :wink:

alexlamsl commented on bin/uglifyjs:
```diff
@@ -173,6 +173,10 @@ function convert_ast(fn) {
 }
 
 function run() {
+    if (program.sourceMap && program.sourceMap.content) {
```
This won't work for `sourceMap.content = "inline"`

b-fuze replied:
Should I add it to the if or should I just take your suggestions above?

alexlamsl replied:
If you don't mind using https://github.com/mishoo/UglifyJS2/pull/3040#issuecomment-377664963 - the clean-up is very useful since the only purpose for `parse_source_map()` was to handle `sourceMap.content`

b-fuze replied:
Okay, thanks

b-fuze commented:
Hmm, I left the function but removed the global variable (didn't want to change more than necessary, probably a bad idea?)

But here are the tests @alexlamsl

alexlamsl commented:
There's a problem with the test - it passes on current `master`.

b-fuze commented:
Okay, I pushed your changes

b-fuze commented:
I think I know why, I'll fix it

b-fuze commented:
@alexlamsl can you check this test against master?

alexlamsl commented:
Unfortunately, your latest test pass with flying colours on `master` :sweat:

alexlamsl commented:
I've tried this based on your test inputs with more success:
```diff
diff --git a/test/input/pr-3040/expect.js b/test/input/pr-3040/expect.js
new file mode 100644
index 0000000..b911f3f
--- /dev/null
+++ b/test/input/pr-3040/expect.js
@@ -0,0 +1,2 @@
+function _toConsumableArray(arr){if(Array.isArray(arr)){for(var i=0,arr2=Array(arr.length);i<arr.length;i++){arr2[i]=arr[i]}return arr2}else{return Array.from(arr)}}var _require=require("bar"),foo=_require.foo;var _require2=require("world"),hello=_require2.hello;foo.x.apply(foo,_toConsumableArray(foo.y(hello.z)));
+//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbImlucHV0Mi5qcyJdLCJuYW1lcyI6WyJyZXF1aXJlIiwiYXJyIl0sIm1hcHBpbmdzIjoiMEpBQWNBLEtBQVFDIiwic291cmNlc0NvbnRlbnQiOlsiY29uc3Qge2Zvb30gPSByZXF1aXJlKFwiYmFyXCIpO1xuY29uc3Qge2hlbGxvfSA9IHJlcXVpcmUoXCJ3b3JsZFwiKTtcblxuZm9vLngoLi4uZm9vLnkoaGVsbG8ueikpO1xuIl19
```
```diff
diff --git a/test/input/pr-3040/input.js b/test/input/pr-3040/input.js
new file mode 100644
index 0000000..6515dcb
--- /dev/null
+++ b/test/input/pr-3040/input.js
@@ -0,0 +1,11 @@
+function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }
+
+var _require = require("bar"),
+    foo = _require.foo;
+
+var _require2 = require("world"),
+    hello = _require2.hello;
+
+foo.x.apply(foo, _toConsumableArray(foo.y(hello.z)));
+
+//# sourceMappingURL=input.js.map
\ No newline at end of file
```
```diff
diff --git a/test/input/pr-3040/input.js.map b/test/input/pr-3040/input.js.map
new file mode 100644
index 0000000..f0883e2
--- /dev/null
+++ b/test/input/pr-3040/input.js.map
@@ -0,0 +1 @@
+{"version":3,"sources":["input2.js"],"names":["require","foo","hello","x","apply","_toConsumableArray","y","z"],"mappings":"kLAAcA,QAAQ,OAAfC,aAAAA,kBACSD,QAAQ,SAAjBE,gBAAAA,MAEPD,IAAIE,EAAJC,MAAAH,IAAAI,mBAASJ,IAAIK,EAAEJ,MAAMK","sourcesContent":["const {foo} = require(\"bar\");\nconst {hello} = require(\"world\");\n\nfoo.x(...foo.y(hello.z));\n"]}
\ No newline at end of file
```
```diff
diff --git a/test/mocha/cli.js b/test/mocha/cli.js
index 10f0465..aac2113 100644
--- a/test/mocha/cli.js
+++ b/test/mocha/cli.js
@@ -1,9 +1,9 @@
 var assert = require("assert");
 var exec = require("child_process").exec;
-var readFileSync = require("fs").readFileSync;
+var fs = require("fs");

 function read(path) {
-    return readFileSync(path, "utf8");
+    return fs.readFileSync(path, "utf8");
 }

 describe("bin/uglifyjs", function () {
@@ -94,6 +94,30 @@ describe("bin/uglifyjs", function () {
             done();
         });
     });
+    it("Should not load source map before finish reading from STDIN", function(done) {
+        var mapFile = "tmp/input-3040.js.map";
+        try {
+            fs.unlinkSync(mapFile);
+        } catch (e) {
+            if (e.code != "ENOENT") throw e;
+        }
+        var command = [
+            uglifyjscmd,
+            "--source-map", "content='" + mapFile + "'",
+            "--source-map", "url=inline"
+        ].join(" ");
+
+        var child = exec(command, function(err, stdout) {
+            if (err) throw err;
+
+            assert.strictEqual(stdout, read("test/input/pr-3040/expect.js"));
+            done();
+        });
+        setTimeout(function() {
+            fs.copyFileSync("test/input/pr-3040/input.js.map", mapFile);
+            child.stdin.end(read("test/input/pr-3040/input.js"));
+        }, 1000);
+    });
     it("Should work with --keep-fnames (mangle only)", function (done) {
         var command = uglifyjscmd + ' test/input/issue-1431/sample.js --keep-fnames -m';
```

b-fuze commented:
What... It fails for someone else and myself...

alexlamsl commented:
> What... It fails for someone else on a mac and myself on linux... It fails however running the whole test suite*. Are you testing it by itself?

I only do `npm test` as well - though I'm on Windows :wink:

b-fuze commented:
Oh dang, I wanted to figure out how to just pipe to the child process directly but thought I wouldn't have the time to look it up, thanks I'll go with your solution

b-fuze commented:
@alexlamsl so I applied your patches but now I'm getting this error when I run `npm test`?

```  157 passing (23s)
  1 failing

  1) bin/uglifyjs Should not load source map before finish reading from STDIN:
     Uncaught Command failed: "/usr/local/bin/node" bin/uglifyjs --source-map content='tmp/input-3040.js.map' --source-map url=inline
Supported options:
  content         null
  filename        null
  includeSources  false
  root            null
  url             null
ERROR: `content=tmp/input-3040.js.map` is not a supported option
    at DefaultsError.get (eval at <anonymous> (/home/b-fuse/Projects/git/UglifyJS2/tools/node.js:21:1), <anonymous>:75:23)
    at fatal (/home/b-fuse/Projects/git/UglifyJS2/bin/uglifyjs:288:53)
    at run (/home/b-fuse/Projects/git/UglifyJS2/bin/uglifyjs:232:9)
    at Socket.<anonymous> (/home/b-fuse/Projects/git/UglifyJS2/bin/uglifyjs:166:9)
    at Socket.emit (events.js:185:15)
    at endReadableNT (_stream_readable.js:1106:12)
    at process._tickCallback (internal/process/next_tick.js:114:19)

```

alexlamsl commented:
I think you've missed the single quotes on my patch - it should read `content='tmp/input-3040.js.map'`

Anyway, I just discovered the test relies on the fact that `tmp/` exists, so you need this to make sure it does:
https://github.com/mishoo/UglifyJS2/blob/41beae4dd71920d2173bb2bc1efeb1231373c2a8/test/fetch.js#L4-L8

alexlamsl commented:
No wait, I see the quotes there - never mind.

Let me have a closer look and report back.

alexlamsl commented:
Right, here's the fix for the problem:
```diff
--- a/test/mocha/cli.js
+++ b/test/mocha/cli.js
@@ -94,6 +94,30 @@ describe("bin/uglifyjs", function () {
     it("Should not load source map before finish reading from STDIN", function(done) {
-        var mapFile = "tmp/input-3040.js.map";
+        var mapFile = "tmp/input.js.map";
+        try {
+            fs.mkdirSync("./tmp");
+        } catch (e) {
+            if (e.code != "EEXIST") throw e;
+        }
         try {
             fs.unlinkSync(mapFile);
         } catch (e) {
             if (e.code != "ENOENT") throw e;
         }
         var command = [
             uglifyjscmd,
-            "--source-map", "content='" + mapFile + "'",
+            "--source-map", "content=" + mapFile + '"',
             "--source-map", "url=inline"
         ].join(" ");
```

alexlamsl commented:
The culprit being that on Linux, those single quotes would be stripped out due to some `bash` syntax pecularities.

So the workaround is to make a name that doesn't require extra quotation so it works across all platforms.

b-fuze commented:
Awesome. I just removed that quote there to make sure it works btw, it fails as expected on master and passes after the patches to uglifyjs. All good now?

```diff
diff --git a/test/mocha/cli.js b/test/mocha/cli.js
index c52d2bb..dc33e11 100644
--- a/test/mocha/cli.js
+++ b/test/mocha/cli.js
@@ -108,7 +108,7 @@ describe("bin/uglifyjs", function () {
         }
         var command = [
             uglifyjscmd,
-            "--source-map", "content=" + mapFile,
+            "--source-map", "content=" + mapFile + '"',
             "--source-map", "url=inline"
         ].join(" ");
 
```

It fails a bit spectacularly on master now, but I don't assume that's an issue```  157 passing (23s)
  1 failing

  1) bin/uglifyjs Should not load source map before finish reading from STDIN:
     Uncaught Command failed: "/usr/local/bin/node" bin/uglifyjs --source-map content=tmp/input.js.map --source-map url=inline
INFO: Using input source map: tmp/input.js.map
Parse error at undefined:undefined,undefined
/home/b-fuse/Projects/git/UglifyJS2/bin/uglifyjs:208
            var lines = files[ex.filename].split(/\r?\n/);
                                           ^

TypeError: Cannot read property 'split' of undefined
    at run (/home/b-fuse/Projects/git/UglifyJS2/bin/uglifyjs:208:44)
    at Socket.<anonymous> (/home/b-fuse/Projects/git/UglifyJS2/bin/uglifyjs:166:9)
    at Socket.emit (events.js:185:15)
    at endReadableNT (_stream_readable.js:1106:12)
    at process._tickCallback (internal/process/next_tick.js:114:19)

```

alexlamsl commented:
Thanks - LGTM!

That spectacular error is due to `bin/uglifyjs` confusing the error from `JSON.parse()` with one from our own parser, so don't worry about it for this PR :wink:

alexlamsl commented on test/mocha/cli.js:
```diff
@@ -94,6 +94,35 @@ describe("bin/uglifyjs", function () {
+            done();
+        });
+        setTimeout(function() {
+            fs.copyFileSync("test/input/pr-3040/input.js.map", mapFile);
```
My bad - didn't realise `fs.copyFile*` is only recently introduced in Node.js 8.5.0

Please change it to the following so it works on all supported versions of Node.js:
```js
fs.writeFileSync(mapFile, read("test/input/pr-3040/input.js.map"));
```

b-fuze replied:
Okay

b-fuze commented:
Okay, pushed

alexlamsl commented on test/mocha/cli.js:
```diff
@@ -94,6 +94,35 @@ describe("bin/uglifyjs", function () {
+        }
+        var command = [
+            uglifyjscmd,
+            "--source-map", "content=" + mapFile + '"',
```
Looks like this trailing quote slipped back in somehow... :sweat_smile:

alexlamsl replied:
I've pushed the fix to your PR, so no further actions required :crossed_fingers:
